duopull/duopull.go

```diff
@@ -0,0 +1,698 @@
+	stackdriverClient *stackdriver.Client
+	datastoreClient   *datastore.Client
+}
+
+// init loads configuration from the environment
+func (c *config) init() error {
+	kms, err := common.NewKMSClient()
+	if err != nil {
+		if debug != debugDuo {
+			log.Fatalf("Could not create kms client. Err: %s", err)
+		} else {
+			kms = &common.KMSClient{}
+		}
+	}
+
+	c.duoAPIHost = os.Getenv("DUOPULL_HOST")
+	c.duoIKey, err = kms.DecryptEnvVar(KEYNAME, "DUOPULL_IKEY")
+	if err != nil {
+		log.Fatalf("Could not decrypt duopull ikey. Err: %s", err)
+	}
+	c.duoSKey, err = kms.DecryptEnvVar(KEYNAME, "DUOPULL_SKEY")
+	if err != nil {
+		log.Fatalf("Could not decrypt duopull skey. Err: %s", err)
+	}
+
+	err = c.validate()
+	if err != nil {
+		return err
+	}
+
+	if debug != debugDuo {
+		c.stackdriverClient, err = stackdriver.NewClient(context.Background(), PROJECT_ID)
+		if err != nil {
+			return err
+		}
+		c.datastoreClient, err = datastore.NewClient(context.Background(), PROJECT_ID)
+		if err != nil {
+			return err
+		}
+	}
+
+	if debug != debugGCP {
+		c.duo = &duoInterface{
+			apiHost: cfg.duoAPIHost,
+			iKey:    cfg.duoIKey,
+			sKey:    cfg.duoSKey,
+		}
+	}
+
+	return nil
+}
+
+// validate verifies the config structure is valid given the operating mode
+func (c *config) validate() error {
+	if debug != debugGCP {
+		if c.duoAPIHost == "" {
+			return fmt.Errorf("DUOPULL_HOST must be set")
+		}
+		if c.duoIKey == "" {
+			return fmt.Errorf("DUOPULL_IKEY must be set")
+		}
+		if c.duoSKey == "" {
+			return fmt.Errorf("DUOPULL_SKEY must be set")
+		}
+	}
+	return nil
+}
+
+// duoInterface is used to generate Authorization headers for requests to the Duo API
+type duoInterface struct {
+	apiHost string
+	iKey    string
+	sKey    string
+}
+
+// getAuthHeader returns an authentication header and date string header for use in a request
+// to the Duo API.
+func (d *duoInterface) getAuthHeader(method, path string, params map[string]string) (string, string) {
+	ds := time.Now().UTC().Format("Mon, 2 Jan 2006 15:04:05 -0700")
+
+	c := []string{
+		ds,
+		strings.ToUpper(method),
+		strings.ToLower(d.apiHost),
+		path,
+	}
+	paramval := url.Values{}
+	for k, v := range params {
+		paramval.Add(k, v)
+	}
+	c = append(c, paramval.Encode())
+	template := strings.Join(c, "\n")
+
+	h := hmac.New(sha1.New, []byte(d.sKey))
+	h.Write([]byte(template))
+
+	auth := fmt.Sprintf("%v:%v", d.iKey, hex.EncodeToString(h.Sum(nil)))
+	return "Basic " + base64.StdEncoding.EncodeToString([]byte(auth)), ds
+}
+
+// logRecords represents a response for a log request from the Duo API.
+//
+// The actual event information is treated as arbitrary JSON. The stat field is
+// specifically included so we can inspect the value to confirm the request was
+// successful.
+//
+// See also https://duo.com/docs/adminapi#api-details
+type logRecords struct {
+	Stat     string        `json:"stat"`
+	Response []interface{} `json:"response"`
+}
+
+type logRecordsV2 struct {
```

Comment:
Do we need both `logRecords` and `logRecordsV2`? Also the version 2 types are missing documentation.